[CASSANDRA-20167][5.0] Reduce memory allocations in miscellaneous places along write path

netudima opened the pull request:
Avoid unnecessary allocations in different places along write path, such as:
- avoid iterator allocations if possible
- handle typical cases (such as a single row, single table writes) more efficiently
- add fast paths for typical scenarios (like pending ranges is empty)
- memorize things which can be computed once

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20167

Commit:
Reduce memory allocations in miscellaneous places along write path

Avoid unnecessary allocations in different places along write path, such as:
- avoid iterator allocations if possible
- handle typical cases (such as a single row, single table writes) more efficiently
- add fast paths for typical scenarios (like pending ranges is empty)
- memorize things which can be computed once

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20167

Commit:
Avoid objects creations if they usually are not needed in a conditional logic
reuse immutable EncodingStats during merging if the result equals to a source
use primitive long + AtomicLongFieldUpdater instead of AtomicLong
DataOutputBuffer - avoid ByteBuffer slicing to create a result byte array

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20167

Commit:
- work with bind variables as with a ImmutableList everywhere to avoid copying of it from a usual collection to an immutable one on every request
- add an optimized SingleTableSinglePartitionUpdatesCollector version for a single row update
- reduce allocations in SingleTableUpdatesCollector for the case when a single partition is updated
- do not create a row builder in PartitionUpdate if it only a single row update

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20167

Commit:
use thread-local pool for BTreeRow Builder in UpdateParameters

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20167

Commit:
- deal with flags in Header and QueryOptions using plain ints and bit logic instead of allocating and using EnumSet to decode/encode every message
- add Flusher.OnFlushCleanup interface and make CQLMessageHandler implement it to avoid a capturing lamda allocation for this::release
- replace payloads.computeIfAbsent with get and put to avoid  a capturing lamda allocation
- avoid Iterator allocation for Flusher.channels on each flush if only V5 connections are used
- iterate over processed ArrayList without an iterator object allocation

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20167

clohfink commented on src/java/org/apache/cassandra/cql3/CQLStatement.java:
```diff
@@ -32,9 +34,9 @@ public interface CQLStatement
-    default List<ColumnSpecification> getBindVariables()
+    default ImmutableList<ColumnSpecification> getBindVariables()
     {
-        return Collections.emptyList();
```
Why switching this to immutable list? emptyList() would have same allocation profile right?

netudima replied:
The problem is with ImmutableList.copyOf(columnSpecs) within OptionsWithColumnSpecifications constructor, we have to copy columnSpecs for every request because in one place plain List is used and in another - ImmutableList

clohfink commented on src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java:
```diff
@@ -270,22 +271,29 @@ protected void onStaticRow(Row row)
             observers.forEach(action);
     }
 
+    private boolean hasObservers()
```
this is to avoid lambda creation correct? as only tables with indexes are expected to have observers?

netudima replied:
Yes, as I see currently the only non-test implementations for observers are related to indexes, and indexes are a rare thing. In the methods we have capturing lamdas (with an externally captured variable value) which are allocated on each invocation, so I am trying to avoid them.

<img width="1294" alt="image" src="https://github.com/user-attachments/assets/9e87750b-7b1f-410f-8051-7f6178df440c" />

netudima replied:
of course, it is not a big saving by itself, but small amounts here and there can add up to something substantial overall.

clohfink commented on src/java/org/apache/cassandra/db/Mutation.java:
```diff
@@ -448,7 +451,7 @@ private Serialization serialization(Mutation mutation, int version)
                     {
                         serializeInternal(PartitionUpdate.serializer, mutation, dob, version);
-                        serialization = new CachedSerialization(dob.toByteArray());
+                        serialization = new CachedSerialization(dob.unsafeToByteArray());
```
if using on heap allocation can we also avoid byte array allocation here with buffer.array()? (buffer.hasArray()). Perhaps provide implementation of CachedSerialization that can use the backed array with offset/length. For larger mutations (ie 64kb-1mb) this can be pretty expensive still

clohfink replied:
err never mind, since using the scratch buffer can't use backing array

netudima replied:
yes, it is reusable, so we have to copy the result

I have one non-implemented idea for this area of logic - initially we invoke serializedSize and then if the result is not too big we invoke the real serialization. In typical cases mutations are small and we can use some cheaper heuristic to decide if it is small to go directy to  real serialization..

netudima replied:
If I find a cheap way to do it - I think I will create a separate improvement for it, it is more about CPU saving (while allocations are also affected a bit)

michaelsembwever commented on src/java/org/apache/cassandra/cql3/VariableSpecifications.java:
```diff
@@ -56,6 +59,17 @@ public List<ColumnSpecification> getBindVariables()
 
+    public ImmutableList<ColumnSpecification> getImmutableBindVariables()
+    {
+        ImmutableList<ColumnSpecification> result = immutableSpecs;
+        if (result == null) // strong syncrhronization is not needed, it is ok if sometimes we create several immutable lists
+        {
+            result = ImmutableList.copyOf(specs);
+            immutableSpecs = result;
+        }
+        return result;
```
why not just
```suggestion
        return immutableSpecs == null ? immutableSpecs  = ImmutableList.copyOf(specs) : immutableSpecs;
```

netudima replied:
immutableSpecs is a volatile variable, in your case we read it 2 times, in my one - just once. It is a slightly better from performance point of view + a less error prone style when you compare and return the same reference which cannot be changed.  It is a kind of typical pattern to deal with volatile variables (and sometimes even with plain fields). You can find it quite a lot in JDK code, for example:
```
Node<K, V>[] tab = this.table;
while(true) {
    int n;
    while(tab == null || (n = tab.length) == 0) {
    tab = this.initTable();
}
```
from: java.util.concurrent.ConcurrentHashMap#putVal

michaelsembwever commented on src/java/org/apache/cassandra/cql3/restrictions/PartitionKeySingleRestrictionSet.java:
```diff
@@ -98,6 +106,11 @@ public List<ByteBuffer> values(QueryOptions options, ClientState state)
         }
+        // memory allocation optimization: to avoid BTreeSet, iterators allocation and use cheaper collections
+        // we handle a single clustering explicitly (we frequently modify a single row)
+        if (builder.buildSize() <= 1) {
+            return toByteBuffers(builder.buildSingle());
+        }
```
```suggestion
        if (builder.buildSize() <= 1)
            return toByteBuffers(builder.buildSingle());

```

Commit:
fix SingleTableSinglePartitionUpdatesCollector logic for the case when a modification statement does not create any mutations

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167
